Firestore/core/test/unit/util/thread_safe_memoizer_test.cc

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2023 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/util/thread_safe_memoizer.h"
+
+#include <thread>  // NOLINT(build/c++11)
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+int ExpensiveFunction() {
+  // Simulate an expensive operation
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  return 42;
+}
+
+TEST(ThreadSafeMemoizerTest, MultiThreadedMemoization) {
```

Comment:
This test ensures that all threads get the same return value from `ExpensiveFunction`, but it does not mean that the function body got executed only once.

I think the following would be better as it ensures (a) all threads get the same result and (b) the function is executed only once:

```cpp
TEST(ThreadSafeMemoizerTest, MultiThreadedMemoization) {
  int global_int = 77;

  auto expensive_lambda = [&]() { // <-- caputes "global_int" by reference
    // Simulate an expensive operation
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    global_int++; // <-- if the lambda gets executed multiple times, threads will see 78, 79, 80, 81, 82.
    return global_int;
  };

  const int num_threads = 5;
  const int expected_result = 78;

  // Create a thread safe memoizer and multiple threads.
  util::ThreadSafeMemoizer<int> memoized_result;
  std::vector<std::thread> threads;

  for (int i = 0; i < num_threads; ++i) {
    threads.emplace_back([&memoized_result, expected_result, &expensive_lambda]() {
      const int& actual_result = memoized_result.memoize(expensive_lambda);

      // Verify that all threads get the same memoized result.
      EXPECT_EQ(actual_result, expected_result);
    });
  }

  for (auto& thread : threads) {
    thread.join();
  }
}
```

Reply:
Good idea. Just make sure to change

```
int global_int = 77;
```

to

```
std::atomic<int> global_int = 77;
```

to avoid a data race

Reply:
the call to `.memoize(expensive_lambda);` should guarantee that `global_int` does not experience a data race. Shouldn't it?

Reply:
Yes, but if the test is verifying that behavior then it should probably assume that there is the possibility that it is broken. If it is indeed broken then it could be a data race. Doesn't hurt to use an atomic just to be safe IMO

Reply:
I see. Agreed. Better to have a test failure than a test crash.